File: src/youtube_automation/utils/image_provider/openai.py
"""OpenAI 画像生成プロバイダー（gpt-image-2 系）。

`openai` SDK の ``images.generate`` / ``images.edit`` を呼ぶ。
``aspect_ratio`` を OpenAI Images API の ``size`` 文字列にマップし、
API キーは `secrets.get_secret("OPENAI_API_KEY")` 経由で取得する。
"""

from __future__ import annotations

import base64
import contextlib
import io
import time
import warnings

from openai import OpenAI

from youtube_automation.utils import cost_tracker
from youtube_automation.utils.exceptions import ConfigError
from youtube_automation.utils.image_provider.base import (
    RETRY_BACKOFF,
    RETRY_MAX,
    ImageGenerationRequest,
    ImageGenerationResult,
)
from youtube_automation.utils.image_provider.composition import log_image_cost, persist_image
from youtube_automation.utils.image_provider.config import (
    OPENAI_SUPPORTED_ASPECT_RATIOS,
    OpenAIConfig,
)
from youtube_automation.utils.secrets import get_secret

# OpenAI Images API が受け付ける size 文字列へのマッピング
_ASPECT_RATIO_TO_SIZE: dict[str, str] = {
    "16:9": "1536x1024",
    "9:16": "1024x1536",
}


class OpenAIImageProvider:
    """OpenAI Images API（gpt-image-2 系）で画像を 1 枚生成して保存する。"""

    name = "openai"
    supported_aspect_ratios: tuple[str, ...] = OPENAI_SUPPORTED_ASPECT_RATIOS

    def __init__(self, config: OpenAIConfig) -> None:
        self._config = config
        # `thinking` は config では受理するが、現在の openai-python SDK は
        # ``images.generate`` / ``images.edit`` に該当 kwarg を渡せない。
        # 設定が効いていると誤認させないため、明示値を持つ場合は警告する。
        # SDK が将来対応したらここを delete し generate() の kwargs に追加する。
        if config.thinking and config.thinking.lower() not in ("off", ""):
            warnings.warn(
                f"OpenAIConfig.thinking={config.thinking!r} は現在 openai-python SDK の "
                "images API に渡せず無視されます",
                stacklevel=2,
            )

    @property
    def pricing_model_id(self) -> str:
        return self._config.model

    def generate(self, req: ImageGenerationRequest) -> ImageGenerationResult:
        """req に従って画像を生成して保存する。"""
        from PIL import Image as PILImage

        if req.aspect_ratio not in _ASPECT_RATIO_TO_SIZE:
            raise ConfigError(
                f"OpenAI image_generation の aspect_ratio={req.aspect_ratio!r} は未対応。"
                f"許容値: {tuple(_ASPECT_RATIO_TO_SIZE.keys())}"
            )

        api_key = get_secret("OPENAI_API_KEY")
        client = OpenAI(api_key=api_key)
        size = _ASPECT_RATIO_TO_SIZE[req.aspect_ratio]
        model = self._config.model
        quality = self._config.quality
        n = self._config.batch
        references = list(req.references)
        save_as_png = req.output_path.suffix.lower() == ".png"

        for attempt in range(RETRY_MAX):
            try:
                ref_label = ""
                if references:
                    names = ", ".join(r.name for r in references)
                    ref_label = f" + 参照画像={names}"
                print(f"  [Submit] モデル={model} size={size} quality={quality}{ref_label}")

                if references:
                    with contextlib.ExitStack() as stack:
                        image_files = [stack.enter_context(ref.open("rb")) for ref in references]
                        response = client.images.edit(
                            model=model,
                            image=image_files,
                            prompt=req.prompt,
                            size=size,
                            quality=quality,
                            n=n,
                        )
                else:
                    response = client.images.generate(
                        model=model,
                        prompt=req.prompt,
                        size=size,
                        quality=quality,
                        n=n,
                    )

                payload = _decode_first_image(response)
                if payload is None:
                    print("  [Retry]  画像なしレスポンス")
                else:
                    pil_image = PILImage.open(io.BytesIO(payload))
                    saved_path = persist_image(pil_image, req.output_path, save_as_png=save_as_png)
                    entry = log_image_cost(
                        model=model,
                        image_size=quality,  # OpenAI は quality で課金階層が決まる → PRICING の by_size key と一致
                        aspect_ratio=req.aspect_ratio,
                        output_file=saved_path,
                        cost_usd=req.cost_per_image_usd,
                        reference_count=len(references),
                    )
                    cost_tracker.print_last_report(entry)
                    return ImageGenerationResult(success=True, saved_path=saved_path)

            except ConfigError:
                # ConfigError は Fail Fast。リトライしない。
                raise
            except Exception as e:  # noqa: BLE001 - SDK 例外は文字列で分岐
                error_msg = str(e)
                print(f"  [Retry]  attempt {attempt + 1}/{RETRY_MAX}: {error_msg[:120]}")

            if attempt < RETRY_MAX - 1:
                backoff = RETRY_BACKOFF[attempt]
                print(f"  [Wait]   {backoff}秒待機...")
                time.sleep(backoff)

        return ImageGenerationResult(success=False, saved_path=None)


def _decode_first_image(response) -> bytes | None:
    """OpenAI Images API レスポンスから先頭画像の生バイト列を取り出す。"""
    data = getattr(response, "data", None) or []
    for item in data:
        b64 = getattr(item, "b64_json", None)
        if b64:
            return base64.b64decode(b64)
    return None

File: packages/core/src/image/openai.ts
// OpenAI 画像生成プロバイダー（gpt-image-2 系。Python
// `utils/image_provider/openai.py` の移植）。
//
// `openai` SDK の `images.generate` / `images.edit` を呼ぶ 1-attempt 契約の
// provider（#959）。リトライ・バックオフ・永続化は service 層（`service.ts` の
// `withRetry` + persist）が所有する。`aspectRatio` を OpenAI Images API の `size`
// 文字列にマップし、未対応比率は SDK を呼ぶ前に `config:` prefix Error で fail fast
// （`defaultShouldRetry` が即 rethrow する）。SDK client は注入され（テストは fake で
// 差し替え）、API キーは env から取得する。
//
// #822 で秘密解決を cli 層 (`packages/cli/lib/secrets.ts`) へ移設したため、core は
// op (1Password) を呼べない（ADR 0002 / oxlint で機械担保）。production default は
// `OPENAI_API_KEY` env を直読みし、未設定なら fail fast する。op fallback が必要な
// 経路は cli/service 層が `createClient` を注入して供給する。
// 名前タグ class (ConfigError 等) は #821 で撤廃済み — `config:` prefix が
// ドメインエラーの単一の真実であり、toServiceError でルーティングされる。

import { basename } from "node:path";

import type { ImageGenerationRequest, ImageProvider } from "./base.ts";
import { OPENAI_SUPPORTED_ASPECT_RATIOS } from "./config.ts";
import type { OpenAIConfig } from "./config.ts";
import { isRecord } from "./internal.ts";
import { readReferenceFiles } from "./references.ts";

// `openai` client の最小シェイプ（注入点の seam）。
interface OpenAIClient {
  images: {
    generate(params: unknown): Promise<unknown>;
    edit(params: unknown): Promise<unknown>;
  };
}

/** OpenAIImageProvider の注入依存。省略時は production default が使われる。 */
export interface OpenAIProviderDeps {
  createClient: () => OpenAIClient | Promise<OpenAIClient>;
}

// OpenAI Images API が受け付ける size 文字列へのマッピング。
const ASPECT_RATIO_TO_SIZE: Record<string, string> = {
  "16:9": "1536x1024",
  "9:16": "1024x1536",
};

// `{ data: [{ b64_json }, ...] }` から先頭の b64_json を持つ要素を decode する。
// b64_json を持たない要素はスキップする（openai.py:140-143）。
const decodeFirstImage = (response: unknown): Uint8Array | null => {
  if (!isRecord(response) || !Array.isArray(response.data)) {
    return null;
  }
  for (const item of response.data) {
    if (isRecord(item) && typeof item.b64_json === "string") {
      return new Uint8Array(Buffer.from(item.b64_json, "base64"));
    }
  }
  return null;
};

const defaultCreateClient = async (): Promise<OpenAIClient> => {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error(
      "config: OPENAI_API_KEY が未設定です。" +
        ".env に設定するか、OpenAIImageProvider に createClient を注入してください"
    );
  }
  const { default: OpenAI } = await import("openai");
  return new OpenAI({ apiKey }) as unknown as OpenAIClient;
};

const defaultDeps = (): OpenAIProviderDeps => ({
  createClient: defaultCreateClient,
});

/** OpenAI Images API（gpt-image-2 系）で画像を 1 attempt 生成し、画像 bytes を返す。 */
export class OpenAIImageProvider implements ImageProvider {
  readonly name = "openai";
  readonly supportedAspectRatios: readonly string[] = [
    ...OPENAI_SUPPORTED_ASPECT_RATIOS,
  ];

  private readonly config: OpenAIConfig;
  private readonly deps: OpenAIProviderDeps;

  // deps はテスト/サービス層が SDK client を差し替えるための注入点。省略時のみ
  // production default を使う（fallback ではなく明示的な DI 既定）。
  constructor(config: OpenAIConfig, deps?: OpenAIProviderDeps) {
    this.config = config;
    this.deps = deps ?? defaultDeps();
  }

  async generate(req: ImageGenerationRequest): Promise<Uint8Array> {
    const size = ASPECT_RATIO_TO_SIZE[req.aspectRatio];
    if (size === undefined) {
      // 未対応比率は client 生成・SDK 呼び出し前に fail fast（リトライしない）。
      throw new Error(
        `config: OpenAI image_generation の aspect_ratio=${JSON.stringify(req.aspectRatio)} は未対応。` +
          `許容値: ${JSON.stringify(Object.keys(ASPECT_RATIO_TO_SIZE))}`
      );
    }

    const client = await this.deps.createClient();
    const references = req.references ?? [];

    // SDK エラーはそのまま伝播させる（リトライ判定は service 側の責務）。
    const response = await this.callApi(client, req, size, references);
    const bytes = decodeFirstImage(response);
    if (bytes) {
      return bytes;
    }
    // b64 decode できないレスポンスは未 prefix Error（service 側で retryable と判定される）。
    throw new Error("openai が画像なしレスポンスを返しました");
  }

  // 参照画像があれば images.edit、なければ images.generate を呼ぶ。
  private callApi(
    client: OpenAIClient,
    req: ImageGenerationRequest,
    size: string,
    references: readonly string[]
  ): Promise<unknown> {
    const base = {
      model: this.config.model,
      n: this.config.batch,
      prompt: req.prompt,
      quality: this.config.quality,
      size,
    };
    if (references.length > 0) {
      const image = readReferenceFiles(references).map(
        ({ bytes, path }) => new File([bytes], basename(path))
      );
      return client.images.edit({ ...base, image });
    }
    return client.images.generate(base);
  }
}
